src/System.Management.Automation/engine/Attributes.cs

```diff
@@ -943,43 +970,100 @@ protected override void ValidateElement(object element)
                         Metadata.ValidateNotNullFailure);
             }
 
-            var o = element as PSObject;
-            if (o != null)
+            if(_rangeKind.HasValue)
             {
-                element = o.BaseObject;
+                // Using dynamic to handle different type of numbers such as
+                // int, double, decimal and float. 
+                dynamic val = element;
```

Comment:
The real problem here is the parameter is not typed - so I think this is by design. If the parameter was typed as `[int]`, then it would have worked as expected.

Based on this, I think it's still reasonable to use `GetCommonType`.